### ISSUE TITLE
Feature request: Transform functions

### ISSUE THREAD

**Issue by ashtuchkin** [heroku/react-refetch#21]
(Split from #17)
What do you think about adding some convenience features?

### Data transform functions
For cases where the data returned is inconvenient to work with directly. Maybe we can reuse `then`, but I think it's not possible now.
```javascript
connect((props) => ({
  data: {
    url: "/api/data",
    transform: ((rawValue) => rawValue.real_data)
  }
}))(Component)
```

**Comment by ashtuchkin** [comment]
Transforming directly in request definition is helpful if you're creating a wrapper to the api, i.e.:
```
// api-helpers.js
function myApiRequest(action, data) {
    return {
        url: `/api/${action}`,
        data: data
        transform: ....
    }
}

// in all views
@connect((props) => {
    project: myApiRequest("get_project")
}))
```

That way, we can handle the api returning data in some hard-to-use, but consistent format.

**Comment by hnordt** [comment]
@ryanbrainard said that we can handle it on `PromiseState.then()`.

I think it's a good approach.

**Comment by ryanbrainard** [comment]
What if we had a kind of "identity request" that would allow a `value` to be provided in the request instead of a `url`? This would allow you to use `then()` on the request side instead of the response side:

```javascript
connect((props) => ({
  data: {
    url: "/api/data",
    then: (rawValue) => { 
      value: rawValue.real_data 
    }
  }
}))(Component)
```

It is basically the same as `transform`, but with one more level of indirection to let it be more composable with things like `catch`, `andThen`, etc. It could even be used standalone at the top-level for static data.

I'm not 100% convinced this is the right way to go and open to suggestions. Thoughts?

cc: @jsullivan

**Comment by ashtuchkin** [comment]
This would work for my case. I like it. It's more general.

1) Do I understand correctly that you need yet another level of indirection, like this:
```es6
connect((props) => ({
  data: {
    url: "/api/data",
    then: (rawValue) => ({ 
      data: {   // <- this one
        value: rawValue.real_data
      }
    })
  }
}))(Component)
```

2) We could go even further and allow raw promises as a request value:
```es6
connect((props) => ({
  data1: "/api/data1",
  data2: { 
    url: "/api/data2",
    then: (rawValue) => {
      data2: new Promise((resolve, reject) => { async_action(resolve); })
    }
  },
  data3: genericAsyncActionReturningPromise(),
  data4: anotherAsyncAction().then((rawValue) => rawValue.real_value)

  // another, less intrusive variant
  data5: {
    promise: asyncAction(),
    then: ...
  }
}))(Component)
```

One use case for that is asynchronous loading of requirejs module that contains code for sub-components. The only challenge I see here is that the evaluation of the function given to `connect` needs to be well-defined (i.e. called only once, etc.). What do you think? I can create a separate issue to discuss this.

**Comment by ryanbrainard** [comment]
1) You don't need to specify the `data` key again with `then`. Just specify the value and it replaces the existing one. If it was `andThen`, you would need to specify the key(s), but it would be strange to overwrite the parent key in `andThen`. It is usually used in cases where you need access to the parent, but also kick off other dependent requests (see https://github.com/heroku/react-refetch#chaining-requests). If you wanted to use `andThen` with an identity request, it would look something like this:

```jsx
connect((props) => ({
  data: {
    url: "/api/data",
    andThen: (rawValue) => ({ 
      realData: { // <-- different name from parent `data`
        value: rawValue.real_data
      }
    })
  }
}))(Component)
```

This would give you inject both `data` and `realData`.

2) Want to avoid that for now. Would transforming the value to `Promise.resolve(value)` work for you?

**Comment by ryanbrainard** [comment]
Closing this to be implemented at identity requests in https://github.com/heroku/react-refetch/issues/33. Re-open if there is more discussion to be had.
